### PR TITLE
/verify spec compliance

### LINKED ISSUE #573: Email Verification: Support application/json flow for GET and POST
per [spec](https://github.com/stormpath/stormpath-framework-spec/blob/master/email-verification.md#get-requests-that-prefer-applicationjson), need to update the existing codebase to support GET and POST application/json type requests and return proper responses.

### PR DESCRIPTION
**Pull request by mzumbadoate**
resolves #573 
this code removes /sendVerificationEmail

### CODE REVIEW

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/VerifyController.java [hunk 1]
```diff
@@ -65,28 +106,148 @@ public boolean isNotAllowedIfAuthenticated() {
...
-        } catch (Exception e) {
-            //safest thing to do if token is invalid or if there is an error (could be illegal access)
-            return new DefaultViewModel(logoutUri).setRedirect(true);
         }
+        catch (Exception e) {
+            if (isJsonPreferred(request, response)) {
+                Map<String,Object> model = errorModelFactory.toError(request, e).toMap();
+                errorModelFactory.toError(request, e);
+                response.setStatus(errorModelFactory.toError(request, e).getStatus());
+                return new DefaultViewModel("stormpathJsonView", model);
```
Comment (dogeared) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/VerifyController.java]: should `stormpathJsonView` be a constant somewhere?

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/VerifyController.java [hunk 2]
```diff
@@ -65,28 +106,148 @@ public boolean isNotAllowedIfAuthenticated() {
...
+            if (autoLogin){
+                final AuthenticationResult result = new TransientAuthenticationResult(account);
+                this.authenticationResultSaver.set(request, response, result);
+
+                Map<String,Object> model = new HashMap<String, Object>();
+                model.put("account", accountModelFactory.toMap(account, Collections.EMPTY_LIST));
+                return new DefaultViewModel("stormpathJsonView", model);
+            }
+            else {
+                return new DefaultViewModel("stormpathJsonView");
```
Comment (dogeared) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/VerifyController.java]: yes, definitely should be a constant

**Comment by dogeared** [comment]
@mzumbadoate - would you please provide some UAT instructions?

**Comment by mzumbadoate** [comment]
UAT
please read https://github.com/stormpath/stormpath-framework-spec/blob/master/email-verification.md

set stormpath.web.verifyEmail.enabled = true

HTML:
    Without sptoken,  verify you'r not logged in, go to /verify, you should see a Send verification Email form.
    Once you enter a valid e-mail you'll receive an e-mail with a sptoken.
    You will be redirected to stormpath.web.verifyEmail.nextUri and append ?status=unverified even if the entered  e-mail is valid or not.
    When you goto /verify?sptoken<SPTOKEN>, using the sptoken received in the Email
        If the sptoken is valid and autoLogin is enabled you will be redirected to stormpath.web.login.enabled.uri.nextUri as a logged in user, cookies or authentication credentials will be created.
        If the sptoken is valid but autoLogin false then you will be redirected to stormpath.web.verifyEmail.nextUri
        if the sptoken is invalid you will be redirected to request a new token but with an error this time.

JSON:
    GET Without sptoken,  status 400 and an error message will be shown.
    do a POST with {"email": "foo@bar.com"} as a parameter, once you enter a valid e-mail you'll receive an e-mail with a sptoken.
        If the email is valid or even invalid you will get 200 OK with no body.
    GET with sptoken as a parameter:
        If the sptoken is valid and autoLogin is enabled you will get a 200 status and logged in credentials.
        If the sptoken is valid but autoLogin false then you will get a status code 200
        if the sptoken is invalid an error from the API, according to the Error Handling specification will be shown.

**Comment by dogeared** [comment]
@mzumbadoate - I resolved the merge conflict and I updated all the relevant code to make "stormpathJsonView" a constant. I will UAT tomorrow morning.

**Comment by mrioan** [comment]
UAT Failed - WIP: Still verifying

* The following GET request should work but it is throwing `sptoken parameter not provided`.
```
http GET http://localhost:8080/verify sptoken==67x7XRjzar3Z6IONqN3Qne Accept:application/json
```

**Comment by mrioan** [comment]
I see that there are no TCK tests for the `verify` endpoint. Please be sure to add them and to assert that they pass before considering this PR ready to be re-reviewed

@mzumbadoate back to you. Thanks

**Comment by mzumbadoate** [comment]
@mrioan I've added the TCK for this endpoint, it is VerifyEmailIT, actually in the TCK branch verify_email_test_to_master, the PR is #294
as we spoke in today's stand up let's leave the internationalization issues for a new ticket in order to merge this branch

**Comment by lhazlewood** [comment]
@mzumbadoate branch conflicts :/

**Comment by mrioan** [comment]
## UAT

* `spring-security-webmvc`: If I send
```
http GET http://localhost:8080/verify sptoken==AAA Accept:application/x-www-form-urlencoded
```
I see this error in the log: `com.stormpath.sdk.servlet.http.UnresolvedMediaTypeException: No MediaType could be resolved for this request ({1})and the configured producesMediaTypes settings ({2}). {3}`

NOTE1: This error had already been previously reported. Please carefully review what the reviewer already reported and fix all the issues before considering the feature to be ready for a new round of review
NOTE2: This is not a i18n issue therefore it should be addressed

**Comment by mzumbadoate** [comment]
@mrioan the correct spec is this
https://github.com/stormpath/stormpath-framework-spec/blob/master/email-verification.md
for the GET method I see
GET Requests That Prefer text/html
and
GET Requests That Prefer application/json

I can't find 'application/x-www-form-urlencoded' in the GET spec

**Comment by mzumbadoate** [comment]
error message fixed to read
com.stormpath.sdk.servlet.http.UnresolvedMediaTypeException: No MediaType could be resolved for this request (application/x-www-form-urlencoded) and the configured producesMediaTypes settings (application/json,text/html). The ContentNegotiationResolver was not able to come up with a valid MediaType.

**Comment by mrioan** [comment]
According to [spec](https://github.com/stormpath/stormpath-framework-spec/blob/master/email-verification.md) a request like this should be supported:

```
http POST http://localhost:8080/verify sptoken==XXX Accept:application/x-www-form-urlencoded
```

However it seems that our Content Negatiation component is not able to support this. I do not think this issue should be blocked by this. I created [this other issue](https://github.com/stormpath/stormpath-sdk-java/issues/781) to work on that.

Approved. Please merge. Thanks @mzumbadoate

**Pull request merged**